tests/ui/errors/issue-116502-span-error.stderr

```diff
@@ -0,0 +1,120 @@
+   |             not allowed in type signatures
+...
+LL |     struct S<T = m!()>(m!(), T)
+   |                  ----  ---- in this macro invocation
+   |                  |
+   |                  in this macro invocation
+LL |     where
+LL |         T: Trait<m!()>;
+   |                  ---- in this macro invocation
+   |
+   = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: use type parameters instead
+   |
+LL ~             U
+LL |
+ ...
+LL |     }
+LL ~     struct S<U>(m!(), T)
+   |
+
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for implementations
+  --> $DIR/issue-116502-span-error.rs:15:13
+   |
+LL |             _
+   |             ^
+   |             |
+   |             not allowed in type signatures
+   |             not allowed in type signatures
+   |             not allowed in type signatures
+...
+LL |     impl<T> std::ops::Index<m!()> for dyn Trait<(m!(), T)>
+   |                             ----                 ---- in this macro invocation
+   |                             |
+   |                             in this macro invocation
+LL |     where
+LL |         T: Trait<m!()>,
+   |                  ---- in this macro invocation
+   |
+   = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: use type parameters instead
+   |
+LL ~             U
+LL |
+ ...
+LL |     std::cell::Cell::<m!()>::new(m!());
+LL ~     impl<T, U> std::ops::Index<m!()> for dyn Trait<(m!(), T)>
+   |
+
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for associated types
+  --> $DIR/issue-116502-span-error.rs:15:13
+   |
+LL |             _
+   |             ^ not allowed in type signatures
+...
+LL |         type Output = m!();
+   |                       ---- in this macro invocation
+   |
+   = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for functions
+  --> $DIR/issue-116502-span-error.rs:15:13
+   |
+LL |             _
+   |             ^
+   |             |
+   |             not allowed in type signatures
+   |             not allowed in type signatures
+...
+LL |         fn index(&self, i: m!()) -> &m!() {
+   |                            ----      ---- in this macro invocation
+   |                            |
+   |                            in this macro invocation
+   |
+   = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: use type parameters instead
+   |
+LL ~             T
+LL |
+ ...
+LL |         type Output = m!();
+LL ~         fn index<T>(&self, i: m!()) -> &m!() {
+   |
+help: try replacing `_` with the type in the corresponding trait method signature
+   |
+LL |             {type error}
```

Comment:
We probably should fix this suggestion anyways

Reply:
Seems the `help` diag you pointed out is another issue.

The root cause in the issue is we inserted duplicated `span` into `HirPlaceholderCollector` here:

https://github.com/rust-lang/rust/blob/a48396984ae7637aa6cbcfad645ef6a48a00f088/compiler/rustc_hir_analysis/src/collect.rs#L129

in the scenario of code like:

```rust
    macro_rules! m {
        () => {
            _         // This is the duplicated span ...
        };
    }
    struct S<T = m!()>(m!(), T)
    where
        T: Trait<m!()>;
```
This PR is trying to remove the duplicated ones, since we're mapping spans with same type `type_name`.

Maybe we need to use `Set` instead of `Vec` in `HirPlaceholderCollector`? I'm not sure about it.